### ISSUE TITLE
BytePointer creates leak

### ISSUE THREAD

**Issue by nickbernz** [bytedeco/javacpp-presets#380]
In BytePointer::BytePointer(ByteBuffer buffer), hasArray() is used to check whether or not buffer is direct. But since it is possible for a direct buffer to return true on that method, this leads to a memory leak. 

```
public BytePointer(ByteBuffer buffer) {
        super(buffer);
        if (buffer != null && buffer.hasArray()) {
            byte[] array = buffer.array();
            allocateArray(array.length - buffer.arrayOffset());
            put(array, buffer.arrayOffset(), array.length - buffer.arrayOffset());
            position(buffer.position());
            limit(buffer.limit());
        }
    }
```

I would suggest to using instead is isDirect().

```
public BytePointer(ByteBuffer buffer) {
        super(buffer);
        if (buffer != null && !buffer.isDirect()) {
            byte[] array = buffer.array();
            allocateArray(array.length - buffer.arrayOffset());
            put(array, buffer.arrayOffset(), array.length - buffer.arrayOffset());
            position(buffer.position());
            limit(buffer.limit());
        }
    }
```

**Comment by saudet** [comment]
Do you have an example of a direct buffer with an array?

There can also be non-direct buffers without array, so it's not safe to call `array()` unless `hasArray()` returns true...

**Comment by nickbernz** [comment]
On my Pixel, all output buffers that come from the AVC decoder (MediaCodec) are direct buffers with array.

Maybe you could check both conditions?

`if (buffer != null && !buffer.isDirect() && buffer.hasArray()) {`

**Comment by saudet** [comment]
Interesting. Yes, that would be a good idea. Could you send a PR for all
the primitive Pointer classes?

**Comment by nickbernz** [comment]
I am sorry, not sure what you are asking for.

**Comment by nickbernz** [comment]
I am sorry, I am not sure what you are asking for.

**Comment by saudet** [comment]
A pull request to send your patch?

**Comment by saudet** [comment]
Please do send a patch to fix this issue! Thanks a lot

**Comment by saudet** [comment]
I've included in JavaCPP 1.3.2 the changes you proposed. Thanks for reporting!
